src/Features/Core/Portable/ConvertForToForEach/AbstractConvertForToForEachCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,504 @@
+
+            // Check in the class/struct hierarchy first.
+            var currentProperty = TryFindMemberInThisOrBaseTypes<IPropertySymbol>(
+                containingType, getEnumeratorReturnType, WellKnownMemberNames.CurrentPropertyName);
+            if (currentProperty != null)
+            {
+                iterationType = currentProperty.Type;
+                return true;
+            }
+
+            // couldn't find .Current on the class/struct.  Check the interface hierarchy.
+            var instantiatedIEnumeratorType = getEnumeratorReturnType.GetAllInterfacesIncludingThis().FirstOrDefault(
+                t => Equals(t.OriginalDefinition, ienumeratorType));
+
+            if (instantiatedIEnumeratorType != null)
+            {
+                iterationType = instantiatedIEnumeratorType.TypeArguments[0];
+                return true;
+            }
+
+            iterationType = default;
+            return false;
+        }
+
+        private async Task<Document> ConvertForToForEachAsync(
+            Document document, TForStatementSyntax forStatement,
+            SyntaxToken iterationVariable, TExpressionSyntax collectionExpression,
+            INamedTypeSymbol containingType, ITypeSymbol collectionType,
+            ITypeSymbol iterationType, CancellationToken cancellationToken)
+        {
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var semanticFacts = document.GetLanguageService<ISemanticFactsService>();
+            var generator = SyntaxGenerator.GetGenerator(document);
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var editor = new SyntaxEditor(root, generator);
+
+            // create a dummy "list[i]" expression.  We'll use this to find all places to replace
+            // in the current for statement.
+            var indexExpression = generator.ElementAccessExpression(
+                collectionExpression, generator.IdentifierName(iterationVariable));
+
+            // See if the first statement in the for loop is of the form:
+            //      var x = list[i]   or
+            //
+            // If so, we'll use those as the iteration variables for the new foreach statement.
+            var (typeNode, foreachIdentifier, declarationStatement) = tryDeconstructInitialDeclaration();
+
+            if (typeNode == null)
+            {
+                // user didn't provide an explicit type.  Check if the index-type of the collection
+                // is different from than .Current type of the enumerator.  If so, add an explicit
+                // type so that the foreach will coerce the types accordingly.
+                var indexerType = GetIndexerType(containingType, collectionType);
+                if (!Equals(indexerType, iterationType))
```

Comment:
Should we check or assert if a valid conversion exists between `indexerType` and `iterationType`. Do we have a unit test to cover this code path?

Reply:
No.  I don't believe so.  i think it's sufficient to just cast when the types are different.